### PR TITLE
Remove binary operator handling through builtins

### PR DESCRIPTION
**Pull request by mistasse**

### CODE REVIEW

File: vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java [hunk 1]
```diff
@@ -110,6 +110,20 @@ public static void defineBuiltins() {
...
+			builtin = Builtin.DEFAULT;
+		}
+
+		Object[] arguments = new OzNode[args.length];
+		for (int i = 0; i < args.length; i++) {
+			arguments[i] = argumentNode(builtin, i, args[i]);
+		}
+
+		return factory.createNode(arguments);
+	}
```
Comment (eregon) [inline vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java:125]: You should reuse this in installBuiltins or it's duplication :p

File: vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java [hunk 2]
```diff
@@ -157,12 +171,16 @@ private static void installBuiltins(String module, List<NodeFactory<? extends Oz
 
 	private static OzNode readArgumentNode(Builtin builtin, int i) {
 		ReadArgumentNode argumentNode = new ReadArgumentNode(i);
+		return argumentNode(builtin, i, argumentNode);
+	}
+
+	private static OzNode argumentNode(Builtin builtin, int i, OzNode arg) {
```
Comment (eregon) [inline vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java]: I propose `transformArgumentNode`

File: vm/src/org/mozartoz/truffle/translator/Translator.java [hunk 3]
```diff
@@ -531,31 +535,35 @@ private Arity buildArity(OzArity arity) {
 	}
 
 	private OzNode translateBinaryOp(String operator, OzNode left, OzNode right) {
-		left = deref(left);
-		right = deref(right);
 		switch (operator) {
+		case "==":
+			return BuiltinsManager.createNodeFromFactory(EqualNodeFactory.getInstance(), left, right);
```
Comment (eregon) [inline vm/src/org/mozartoz/truffle/translator/Translator.java]: The lines are a bit long, maybe we can simply have a helper function `createBinaryNode` or so.

File: vm/src/org/mozartoz/truffle/translator/Translator.java [hunk 4]
```diff
@@ -580,6 +588,10 @@ private OzNode deref(OzNode node) {
 		return DerefNode.create(node);
 	}
 
+	private OzNode tryDeref(OzNode node) {
+		return DerefIfBoundNode.create(node);
+	}
```
Comment (eregon) [inline vm/src/org/mozartoz/truffle/translator/Translator.java]: Dead code

File: vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java (hunk 1 above)
Comment (mistasse) [reply vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java:125]: Actually, I modified readArgumentNode in order to use it. But I'm going to inline it directly in installBuiltins.

Comment (eregon) [reply vm/src/org/mozartoz/truffle/nodes/builtins/BuiltinsManager.java:125]: I meant reusing createNodeFromFactory in installBuiltins

File: vm/src/org/mozartoz/truffle/translator/Translator.java [hunk 5]
```diff
@@ -531,36 +535,44 @@ private Arity buildArity(OzArity arity) {
...
+			return createBinaryOp(DotNodeFactory.getInstance(), left, right);
+		case ":=":
+			return createBinaryOp(CatExchangeNodeFactory.getInstance(), left, right);
 		default:
 			throw unknown("operator", operator);
 		}
 	}
 
+	private static OzNode createBinaryOp(NodeFactory<? extends OzNode> factory, OzNode left, OzNode right) {
+		return BuiltinsManager.createNodeFromFactory(factory, new OzNode[] { left, right });
```
Comment (eregon) [inline vm/src/org/mozartoz/truffle/translator/Translator.java:573]: Just as a note: you could keep the varargs signature, it also accepts an OzNode[] as a single final argument in Java.

**Comment by eregon** [comment]
Very nice, thanks for the improvement!

**Pull request merged**